cpp/include/cugraph/graph_mask.hpp

```diff
@@ -0,0 +1,382 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/core/device_span.hpp>
+#include <raft/core/handle.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <cstdint>
+#include <limits>
+#include <optional>
+#include <vector>
+
+namespace cugraph {
+
+/**
+ * Compile-time fast lookup of log2(num_bits(mask_t)) to eliminate
+ * the log2 computation for powers of 2.
+ * @tparam mask_t
+ */
+template <typename mask_t>
+__host__ __device__ constexpr int log_bits()
+{
+  switch (std::numeric_limits<mask_t>::digits) {
+    case 8: return 3;
+    case 16: return 4;
+    case 32: return 5;
+    case 64: return 6;
+    default: return log2(std::numeric_limits<mask_t>::digits);
+  }
+}
+
+/**
+ * Uses bit-shifting to perform a fast mod operation. This
+ * is used to compute the index of a specific bit
+ * @tparam mask_t
+ * @tparam T
+ */
+template <typename mask_t, typename T>
+__host__ __device__ int bit_mod(T numer)
```

Comment:
I assume `numer` here is a typo of `number`.

Reply:
It's actually the `numerator` of the division for the mod but i can change it to `number` if that's what you prefer

Reply:
I see, then, this is not a big issue, but I slightly prefer either `number` or `numerator` than `numer`. It is little difficult to infer `numerator` from `numer`.